03 merge asf senedd repo

crispy-wonton opened the pull request:
Fixes #3 and #7

Changes:
- add relevant code to produce figures from asf_senedd_response repo into this repo
- update README expected outputs section
- update config variable name to config_file in getters/get_data.py
- update time_series_comparison function in pipeline/plotting.py to improve the way default values are determined for domain min and max params
- add supp_data arg to specify which supplementary data folder to use

# Instructions for Reviewer
- I have labelled code where I've copied it from asf_senedd_response. This code shouldnt need a detailed review as it's from the dev branch and has already been reviewed
- please could you review the processing of dfs before they are used to generate graphs to ensure there's no processing mistakes
- please could you review the time_series_comparison function in pipeline/plotting.py to check changes i've made make sense and work
- please could you run `python asf_welsh_energy_consultation/analysis/produce_plots.py --local_data_dir <YOUR_LOCAL_DIR>` from this branch to ensure it still works and check you are getting all the graphs as outlined in the updated README

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

crispy-wonton committed:
update README

improve and correct set up instructions and update output file extensions to html.

crispy-wonton committed:
update local_data_dir in getters

update local_data_dir to use relative file path from config base.yaml so config file can be updated by users rather than editing scripts.

crispy-wonton committed:
update requirements.txt with versions

crispy-wonton committed:
change outputs to .html files

change output graphs from produce_plots.py from .png to .html files to avoid altair CalledProcessError

crispy-wonton committed:
fix introduced typo in requirements.txt

crispy-wonton committed:
automate all data downloads/loading in script

- add argparser to allow user to specify local_data_dir, epc-, and mcs-batch
- add function to load MCS data directly from S3 bucket
- add function to check for EPC data locally and download from S3 if not located
- update existing functions to work with above
- add global parameters for EPC processing version to base.yaml config file

crispy-wonton committed:
update README

update instructions in README to explain new way of running script and to record batches for historical analyses

crispy-wonton committed:
update requirements.txt with argpase

crispy-wonton committed:
merge asf_senedd_response wales_analysis.py

add script to generate stats and charts from asf_senedd_response into produce_plots.py

crispy-wonton committed:
add global parameters to config file

crispy-wonton committed:
merge asf_senedd_response getters from loading.py

merge getters from loading.py in asf_senedd_response into get_data.py script

crispy-wonton committed:
update function name

update EPC getter function name to match updates in get_data.py

crispy-wonton committed:
add new files from asf_senedd_response repo

add files:
- translation_config for welsh translations
- plotting.py for plotting functions
- augmenting.py for data processing and augmentation

crispy-wonton committed:
resolve merge conflits - merge branch 'dev' into 03_merge_asf_senedd_repo

crispy-wonton commented on asf_welsh_energy_consultation/analysis/produce_plots.py:
```diff
@@ -21,57 +28,6 @@
-    height=300,
-):
-    """Generic function for plotting a line chart by category (represented by color_var).
-
-    Args:
-        data (pd.DataFrame): Base data. Needs to be structured as a column of consecutive dates,
-            a column indicating categories and a column with cumulative values.
-        title (str/list): Chart title.
-        y_var (str): y variable.
-        y_title (str): y axis title.
-        color_var (str): Variable to split by.
-        x_var (str, optional): x variable. Defaults to "date:T".
-        x_title (str, optional): x axis title. Defaults to "Date".
-        domain_min (str, optional): x axis minimum. Defaults to "2015-01-01".
-        domain_max (str, optional): x axis maximum. Defaults to "2023-01-01".
-        width (int, optional): Chart width. Defaults to 600.
-        height (int, optional): Chart height. Defaults to 300.
-
-    Returns:
-        alt.Chart: Base altair chart.
-    """
-    chart = (
-        alt.Chart(
-            data,
-            title=title,
-        )
-        .mark_line()
-        .encode(
-            x=alt.X(
-                x_var, title=x_title, scale=alt.Scale(domain=[domain_min, domain_max])
-            ),
-            y=alt.Y(y_var, title=y_title),
-            color=color_var,
-        )
-        .properties(width=width, height=height)
-    )
-
-    return chart
-
-
```
Moved to pipeline/plotting

crispy-wonton commented on asf_welsh_energy_consultation/analysis/produce_plots.py:
```diff
@@ -213,3 +169,195 @@ def time_series_comparison(
+    # EPC, private sector with HPs
+    proportions_bar_chart(
+        wales_hp.loc[wales_hp.TENURE.isin(["Perchen-feddiannaeth", "Rhentu preifat"])],
+        "CURRENT_ENERGY_RATING",
+        [
+            "Ffig. 7: Sgoriau EPC eiddo perchen-feddiannaeth a",
+            "rhentu preifat Cymru sydd â phympiau gwres",
+        ],
+        "Sgôr effeithlonrwydd ynni",
+        "Canran yr eiddo",
+        filename="epc_hp_private_welsh",
+        x_type="other",
+        language="welsh",
+    )
+
+    # EPCs, private sector with retrofitted HPs
+    proportions_bar_chart(
+        wales_hp.loc[
+            wales_hp.TENURE.isin(["Perchen-feddiannaeth", "Rhentu preifat"])
+            & (wales_hp.CONSTRUCTION_AGE_BAND != "2007 ymlaen")
+        ],
+        "CURRENT_ENERGY_RATING",
+        [
+            "Ffig. 8: Sgoriau EPC eiddo perchen-feddiannaeth a rhentu prifat",
+            "Cymru sydd â phympiau gwres, a adeiladwyd cyn 2007",
+        ],
+        "Sgôr effeithlonrwydd ynni",
+        "Canran yr eiddo",
+        filename="epc_hp_private_retrofit_welsh",
+        x_type="other",
+        language="welsh",
+    )
+
+    # Ages and EPC ratings
+    age_prop_chart(
+        age_data,
+        "Ffig. 9: Bandiau oedran adeiladu ac effeithlonrwydd ynni",
+        "age_prop_welsh",
+        language="welsh",
+    )
```
Copied from asf_senedd_response repo

crispy-wonton commented on asf_welsh_energy_consultation/config/translation_config.py:
Copied from asf_senedd_response repo

crispy-wonton commented on asf_welsh_energy_consultation/getters/get_data.py:
```diff
@@ -351,3 +355,77 @@ def get_electric_tenure():
+                "TRANSACTION_TYPE",
+                "HP_INSTALLED",
+            ],
+        )
+
+        wales_epc.TENURE = wales_epc.TENURE.replace(
+            {
+                "owner-occupied": "Owner-occupied",
+                "rental (social)": "Socially rented",
+                "rental (private)": "Privately rented",
+                "unknown": "Unknown",
+            }
+        )
+        # if CONSTRUCTION_AGE_BAND is unknown and TRANSACTION_TYPE is new dwelling,
+        # assume construction age is >2007 because EPCs started in 2008
+        wales_epc["CONSTRUCTION_AGE_BAND"].loc[
+            (wales_epc.CONSTRUCTION_AGE_BAND == "unknown")
+            & (wales_epc.TRANSACTION_TYPE == "new dwelling")
+        ] = "2007 onwards"
+
+        if not os.path.isdir(input_data_path):
+            os.makedirs(input_data_path)
+
+        wales_epc.to_csv(input_data_path + wales_epc_path)
+
+    return wales_epc
+
+
+def load_wales_hp(wales_epc):
+    """Load Welsh EPC data filtered to properties with heat pumps.
+
+    Args:
+        wales_epc (pd.DataFrame): Wales EPC data.
+
+    Returns:
+        pd.DataFrame: EPC data filtered to properties with heat pumps.
+    """
+    wales_hp = wales_epc.loc[wales_epc.HP_INSTALLED].reset_index(drop=True)
+
+    return wales_hp
```
Copied from asf_senedd_response repo

crispy-wonton commented on asf_welsh_energy_consultation/pipeline/augmenting.py:
Copied from asf_senedd_response repo

crispy-wonton commented on asf_welsh_energy_consultation/pipeline/plotting.py:
```diff
@@ -0,0 +1,268 @@
+    domain_max="2023-01-01",
+    width=600,
+    height=300,
+):
+    """Generic function for plotting a line chart by category (represented by color_var).
+
+    Args:
+        data (pd.DataFrame): Base data. Needs to be structured as a column of consecutive dates,
+            a column indicating categories and a column with cumulative values.
+        title (str/list): Chart title.
+        y_var (str): y variable.
+        y_title (str): y-axis title.
+        color_var (str): Variable to split by.
+        x_var (str, optional): x variable. Defaults to "date:T".
+        x_title (str, optional): x-axis title. Defaults to "Date".
+        domain_min (str, optional): x-axis minimum. Defaults to "2015-01-01".
+        domain_max (str, optional): x-axis maximum. Defaults to "2023-01-01".
+        width (int, optional): Chart width. Defaults to 600.
+        height (int, optional): Chart height. Defaults to 300.
+
+    Returns:
+        alt.Chart: Base altair chart.
+    """
+    chart = (
+        alt.Chart(
+            data,
+            title=title,
+        )
+        .mark_line()
+        .encode(
+            x=alt.X(
+                x_var, title=x_title, scale=alt.Scale(domain=[domain_min, domain_max])
+            ),
+            y=alt.Y(y_var, title=y_title),
+            color=color_var,
+        )
+        .properties(width=width, height=height)
+    )
+
+    return chart
```
Moved from produce_plots

crispy-wonton commented on asf_welsh_energy_consultation/pipeline/plotting.py:
Mostly copied from asf_senedd_response repo

crispy-wonton commented on asf_welsh_energy_consultation/pipeline/process_data.py:
```diff
@@ -102,7 +102,7 @@ def cumsums_by_variable(variable, new_var_name, data=enhanced_mcs):
 # PROCESSING EPC
 
-wales_epc = get_wales_epc()
+wales_epc = get_wales_processed_epc()
```
Updated for clarity on version

crispy-wonton committed:
update config variable name in __init__.py

update to prevent module not subscriptable error

crispy-wonton committed:
fix minor errors in produce_plots

crispy-wonton committed:
use new config_file variable name

crispy-wonton committed:
update domain min and max variables in time_series_comparison

improve selection of min and max dates

crispy-wonton commented on asf_welsh_energy_consultation/pipeline/plotting.py:
```diff
@@ -0,0 +1,277 @@
+    Returns:
+        alt.Chart: Base altair chart.
+    """
+    if domain_min is None:
+        domain_min = config_file["plots"]["time_series_min_default"]
+        logger.info(f"Time series comparison using {domain_min} as min date")
+    if domain_max is None:
+        domain_max = data.date.max()
+        logger.info(f"Time series comparison using {domain_max} as max date")
```
this part is new

crispy-wonton committed:
update README with new output files

crispy-wonton committed:
update plot max dates for newest EPC/MCS batches

crispy-wonton committed:
add supp_data arg

allows user to specify which directory with supplementary data to use, meaning new supplementary data can be added to analysis as it's updated

crispy-wonton committed:
delete formatting file and merge augmenting with process_data.py

crispy-wonton commented on asf_welsh_energy_consultation/__init__.py:
```diff
@@ -32,4 +32,4 @@ def get_yaml_config(file_path: Path) -> Optional[dict]:
 # base/global config
 _base_config_path = Path(__file__).parent.resolve() / "config/base.yaml"
-config = get_yaml_config(_base_config_path)
+config_file = get_yaml_config(_base_config_path)
```
changed to avoid conflict errors when importing multiple configs

crispy-wonton committed:
add logging and improve how default domain min and max is determined

crispy-wonton committed:
update import statements to remove import *

sofiapinto commented on asf_welsh_energy_consultation/analysis/produce_plots.py:
```diff
@@ -129,7 +91,7 @@ def time_series_comparison(
                 "year",
                 title="Year",
-                scale=alt.Scale(domain=["2007-07-01", "2022-06-01"]),
+                scale=alt.Scale(domain=["2007-07-01", "2023-01-01"]),
```
I think I asked Chris about this before, but it might be nice to have max date based on the input data? Otherwise we will have to change this in future consultations.

sofiapinto commented on asf_welsh_energy_consultation/analysis/produce_plots.py:
```diff
@@ -167,8 +129,8 @@ def time_series_comparison(
 
-    ret = get_mcs_retrofits()
-    ret_cumsums = cumsums_by_variable("country", "wales_col", data=ret)
+    ret = process_data.get_mcs_retrofits()
```
shall we call it `retrofits` or `mcs_retrofits` instead of `ret`?

sofiapinto commented on asf_welsh_energy_consultation/analysis/produce_plots.py:
```diff
@@ -200,7 +162,7 @@ def time_series_comparison(
             electric_tenure,
             title="Fig. 2: Properties in Wales with only electric heating, split by tenure (N = "
-            + format_number(N)
+            + "{:,}".format(N)
```
have never seen this `{:,}` - cool!

crispy-wonton replied:
yeh! found it online! :)

sofiapinto commented on asf_welsh_energy_consultation/analysis/produce_plots.py:
```diff
@@ -213,3 +175,200 @@ def time_series_comparison(
+
+    # English plots
+
+    # Key statistics
+    print("Number of heat pumps:", len(wales_hp))
+    print("Number of properties in EPC:", len(wales_df))
+    print(
+        "Estimated percentage of properties with a heat pump:",
+        "{:.2%}".format(len(wales_hp) / len(wales_df)),
+    )
```
a little bit picky, but shall we rename this `produce_plots_and_statistics.py` since it's not just plots?

sofiapinto replied:
also maybe we could use logger instead of print, just to use logger consistently across the repo.

crispy-wonton replied:
I've saved these print outputs into a txt file instead for easier use of the stats.

sofiapinto commented on asf_welsh_energy_consultation/analysis/produce_plots.py:
```diff
@@ -213,3 +175,200 @@ def time_series_comparison(
+
+    ## Welsh plots
+
+    welsh_replacements = {
+        "TENURE": {
+            "Owner-occupied": "Perchen-feddiannaeth",
+            "Socially rented": "Rhentu cymdeithasol",
+            "Privately rented": "Rhentu preifat",
+            "Unknown": "Anhysbys",
+        },
+        "CONSTRUCTION_AGE_BAND": {
+            "England and Wales: before 1900": "Cyn 1900",
+            "Pre-1900": "Cyn 1900",
+            "2007 onwards": "2007 ymlaen",
+            "unknown": "Anhysbys",
+        },
+    }
```
any reason for this not to be in `translation_config.py`?

crispy-wonton replied:
nope, moved it :d

sofiapinto commented on asf_welsh_energy_consultation/getters/get_data.py:
```diff
@@ -78,10 +74,28 @@ def get_args():
-    return parser.parse_args()
+    args = parser.parse_args()
+
+    if args.supp_data == "newest":
+        subdirs = [subdir for subdir in os.listdir("inputs")]
+        args.supp_data = max(subdirs)
+
+    return args
```
what happens if not `newest`? there's no else statement

crispy-wonton replied:
it uses the user defined arg which is the folder name. this part basically just converts 'newest' to the folder name instead.

sofiapinto commented on asf_welsh_energy_consultation/config/base.yaml:
```diff
@@ -1,3 +1,5 @@
 epc_data_config:
   epc_processing_version: "preprocessed"
```
is this still in use?

crispy-wonton replied:
no - removed

sofiapinto commented on asf_welsh_energy_consultation/pipeline/plotting.py:
```diff
@@ -0,0 +1,275 @@
+    if not os.path.isdir(file_path):
+        os.makedirs(file_path)
+
+setup_theme()
```
nothing related to the code itself but I noticed your figures in the GoogleDrive don't have the font we usually use for Nesta plots, so it might be that you need to install it in your computer for the plots to be generated with it: it's the Averta font.

sofiapinto replied:
Maybe you could add a comment above explaining what this is doing?

sofiapinto commented on asf_welsh_energy_consultation/pipeline/process_data.py:
```diff
@@ -17,26 +18,34 @@ def get_enhanced_mcs():
+        logger.warning(
+            f"{mcs.country.isna().sum()} MCS installation records have no country match."
+            f"Potential loss of data when filtering for Wales."
+        )
```
I love these warnings! Very insightful to have them. Would just add either a space of a \n line break - the "match.Potential" on the console freaks me out because it looks like code 😂😂😂
```suggestion
        logger.warning(
            f"{mcs.country.isna().sum()} MCS installation records have no country match. "
            f"Potential loss of data when filtering for Wales."
        )
```

sofiapinto commented on asf_welsh_energy_consultation/pipeline/plotting.py:
```diff
@@ -0,0 +1,275 @@
+
+    plt.savefig(fig_output_path[language] + filename + ".png", bbox_inches="tight")
+
+    print("Saved: " + filename + ".png")
```
maybe we could have this for all images? passing the full path instead of the filename might be helpful for users to go to the directory directly. :) also logger instead of print?

crispy-wonton replied:
done

sofiapinto commented on asf_welsh_energy_consultation/pipeline/process_data.py:
```diff
@@ -5,7 +5,8 @@
 import pandas as pd
 
-from asf_welsh_energy_consultation.getters.get_data import *
+from asf_welsh_energy_consultation.getters import get_data
```
i like your approach :) this way you know where the functions are coming from

sofiapinto commented on asf_welsh_energy_consultation/pipeline/process_data.py:
```diff
@@ -5,7 +5,8 @@
 
-from asf_welsh_energy_consultation.getters.get_data import *
+from asf_welsh_energy_consultation.getters import get_data
+from asf_core_data.getters.data_getters import logger
```
I am a bit of a noob when it comes to logging but couldn't we do the following instead?

```
import logging
logger = logging.getLogger(__name__)
```

crispy-wonton replied:
yes you're right, that's much better. i've updated it

sofiapinto commented on asf_welsh_energy_consultation/pipeline/process_data.py:
```diff
@@ -17,26 +18,34 @@ def get_enhanced_mcs():
+    # There will be records with no match
     # Some will be new postcodes (new build developments)
     # and some may be expired postcodes
     # In future, implement new solution that uses outward codes
```
maybe remove this comment? or maybe update it? from what I remember from our conversation, this is not necessarily a good solution because the outward codes might be either from England or Wales.
Also might be worth adding some info about the percentage in which this happens (around 1% right?). Just so that in future, if someone looks at this they don't need to go check the numbers again.

crispy-wonton replied:
yes good point, deleted now. I have added logging to tell the user how many rows they might lose which is what the logging warning above this code snippet will show :)

sofiapinto commented on README.md:
```diff
@@ -34,6 +33,14 @@ The script should generate the following six plots which will be saved in your l
+- `epc_hp_private[_welsh].html`
+- `hp_tenure[_welsh].html`
+
 ## Skeleton folder structure
```
does the skeleton folder structure need to be updated?

sofiapinto commented on asf_welsh_energy_consultation/getters/get_data.py:
```diff
@@ -51,6 +40,13 @@ def create_argparser():
 
+    parser.add_argument(
+        "--supp_data",
+        help="Name of directory where supplementary data is stored",
```
might be helpful to mention two things in this help: 1) say something about the name format, so that people can use something that's not the "newest" and 2) say that the default is newest

sofiapinto commented on asf_welsh_energy_consultation/getters/get_data.py:
```diff
@@ -4,12 +4,13 @@
 
 from asf_welsh_energy_consultation import PROJECT_DIR
-from asf_welsh_energy_consultation import config
+from asf_welsh_energy_consultation import config_file
```
I don't think `config_file` is being used in this script, so I think we can remove this line

sofiapinto commented on asf_welsh_energy_consultation/getters/get_data.py:
Can't add this comment to the specific line because it's not new/changed code, but line 331: mentions 2021 census data specifically. You might want to remove the year mention for it to be correct for future data?

sofiapinto commented on asf_welsh_energy_consultation/getters/get_data.py:
```diff
@@ -351,3 +367,81 @@ def get_electric_tenure():
+        )
+        # if CONSTRUCTION_AGE_BAND is unknown and TRANSACTION_TYPE is new dwelling,
+        # assume construction age is >2007 because EPCs started in 2008
+        wales_epc["CONSTRUCTION_AGE_BAND"].loc[
+            (wales_epc.CONSTRUCTION_AGE_BAND == "unknown")
+            & (wales_epc.TRANSACTION_TYPE == "new dwelling")
+        ] = "2007 onwards"
```
Note that for new versions this might not change anything (due to this step being added to ASF core data), but it does need to be here to work with older versions of the data :)

crispy-wonton replied:
added a comment

sofiapinto commented on asf_welsh_energy_consultation/pipeline/process_data.py:
Again, can't comment on the specific lines in this cases but:
- regarding `get_wales_epc_new()` line 117-136: we should probably be checking if these new dwellings are actually new dwellings, as in if there’s no previous inspection for the same home, i.e. the "new building" record is the first record for the property. (Since we know that new 6% of properties are erroneously tagged as new dwellings - either that or the UPRNs are wrong)
- `get_wales_epc_new()` function: A better name for this function would be `get_wales_new_builds_epc()`
- Line 139: `get_new_hp_counts()` -> new suggested name: `get_new_builds_hp_counts()`
- Line 207: maybe we should check the % of EPC UPRNs matched to the same installation - since we know that sometimes happens that the same EPC is matched to multiple MCS records. Just to check if this is a high value! If not, we can probably keep this as is.

crispy-wonton replied:
@ point 1: I've added a new function called `correct_new_dwelling_labels` which will change all the 'new dwelling' labels after the first inspection to 'unknown'

sofiapinto commented on README.md:
```diff
@@ -18,7 +18,6 @@ The remainder of the charts in the response can be produced from code in the rep
   - Install packages listed in `requirements.txt`
 - Activate conda environment: `conda activate asf_welsh_energy_consultation`
 - Run `make inputs-pull` to pull the zipped static data from S3 and put it in `/inputs`
```
is this what the inputs folder is supposed to look like at the moment after running `make inputs-pull`? 
<img width="217" alt="image" src="https://github.com/nestauk/asf_welsh_energy_consultation/assets/70645959/544f6cd1-455d-42bb-9b8b-4da65cc5c5a4">

sofiapinto replied:
why do we have a folder for data and a folder for data_202310?

crispy-wonton replied:
yes it's meant to look like that but good point I was meant to ask if you are able to update the name in S3. 'data' contains supplementary data for the April 2023 analysis and 'data_202310' contains supplementary data for this October 2023 analysis. I dont have permissions to rename the 'data' S3 folder to 'data_202304' - do you? If not, I can create a new folder with the correct name and delete the old one. Every time someone updates the analysis, they should upload all their supplementary data (if updated) to a new folder with the year and month. I have added this to the README on [branch 08](https://github.com/nestauk/asf_welsh_energy_consultation/tree/08_add_map)

sofiapinto reviewed (commented):
Hey Roisin @crispy-wonton,

Great work! 👯 thank you

I re-ran `make inputs-pull`  and `python asf_welsh_energy_consultation/analysis/produce_plots.py --local_data_dir <YOUR_LOCAL_DIR>`. All ran well!

I've looked at the code, including parts of the processing code that haven't been changed in this PR - to make sure the processing is fine. Left a couple of comments, but almost all of them are just suggestions, so please feel free to ignore them if they don't make sense to you.

I'd still like to take a better look at some parts of the code + the generated plots in more detail (including thinking if the numbers make sense, as we discussed earlier this week). I can do that on Monday first thing in the morning! Sorry didn't manage to do it today.

Also: You might want to add a line at the top of the README [in the other repo](https://github.com/nestauk/asf_senedd_response) pointing people to this one saying that all code is now here :)

crispy-wonton committed:
make changes requested in review
- update README with correct file structure and new output
- move translation dicts to translation_config.py
- write summary stats into output .txt file instead of printing to console
- correct new dwelling labels
- update function names and documentation for clarity

crispy-wonton committed:
update documentation

crispy-wonton committed:
make improvements suggested in code review

- use logging instead of print
- log info for all charts when saved
- generalise calculation of max date for graphs
- remove unused config

crispy-wonton committed:
remove duplicate UPRNs - indicates multiple MCS installations joined to one EPC record

crispy-wonton committed:
update get_mcs_and_joined_data function to allow passing epc_version param

crispy-wonton committed:
update postcode df to reduce row loss when adding country col to MCS data

also change graphs to nesta blue

sofiapinto commented on README.md:
```diff
@@ -41,20 +41,26 @@ It should generate a further 10 plots, five in English and five in Welsh, saved
 │  ├─ unused_processing.py - unused processing functions from August '22
 inputs/
-├─ data/ - data files, a mixture of csv, xlsx and ods
+├─ data_[YYYYMM]/ - data files, a mixture of csv, xlsx and ods
```
not too important: but i think we're missing `data`. as each YYYYMM folder is under `data`

sofiapinto commented on asf_welsh_energy_consultation/analysis/produce_plots_and_stats.py:
```diff
@@ -91,7 +91,7 @@
                 "year",
                 title="Year",
-                scale=alt.Scale(domain=["2007-07-01", "2023-01-01"]),
+                scale=alt.Scale(domain=["2007-07-01", f"{max_date.year}-07-01"]),
```
this is always up to month 7 - any reason why? or should the month also be defined by the data?

crispy-wonton replied:
yes this will always be the same because this graph only takes data for each full year and not partial years. This domain just helps to create a good margin on the right side of chart

sofiapinto commented on asf_welsh_energy_consultation/analysis/produce_plots_and_stats.py:
```diff
@@ -202,23 +212,27 @@
+    epc_c_wall_roof = f"\n\nNumber of EPC C+ properties with good or very good wall and roof insulation: {len(epc_c_or_above_and_good_walls_and_roof)}\n"
+    epc_c_wall_roof_proportion = f"As a proportion of properties in EPC: {len(epc_c_or_above_and_good_walls_and_roof) / len(wales_df)}"
+
+    with open(os.path.join(output_folder, "stats.txt"), "w") as stats_txt:
+        stats_txt.writelines(
+            [
+                intro,
+                total_hp,
+                total_epc,
+                hp_perc,
+                tenure_value_counts,
+                epc_c_wall,
+                epc_c_wall_proportion,
+                epc_c_wall_roof,
+                epc_c_wall_roof_proportion,
+            ]
+        )
```
nice :)

sofiapinto commented on asf_welsh_energy_consultation/pipeline/process_data.py:
```diff
@@ -114,14 +114,38 @@ def cumsums_by_variable(variable, new_var_name, data=enhanced_mcs):
+    Returns:
+        pd.DataFrame: Wales EPC certificates
+    """
+    wales_epc["rank"] = wales_epc.groupby("BUILDING_REFERENCE_NUMBER")[
+        "INSPECTION_DATE"
+    ].rank("dense")
```
i think we need to use UPRN here, as 1 building might have multiple flats. so if we use `BUILDING_REFERENCE_NUMBER` the results will be wrong
```suggestion
    wales_epc["rank"] = wales_epc.groupby("UPRN")[
        "INSPECTION_DATE"
    ].rank("dense")
```

crispy-wonton replied:
Great point thank you! Got them mixed up here.

sofiapinto commented on asf_welsh_energy_consultation/pipeline/process_data.py:
```diff
@@ -204,13 +229,38 @@ def get_new_hp_cumsums():
     return new_hps_cumsums
 
 
+def identify_mcs_with_multiple_epc():
```
amazing! :) thanks for adding this

sofiapinto commented on README.md:
```diff
@@ -18,7 +18,7 @@ The remainder of the charts in the response can be produced from code in the rep
 - Activate conda environment: `conda activate asf_welsh_energy_consultation`
 - Run `make inputs-pull` to pull the zipped static data from S3 and put it in `/inputs`
-- Run `python asf_welsh_energy_consultation/analysis/produce_plots.py --local_data_dir <YOUR_LOCAL_DIR>`. You need to specify the path to the local
+- Run `python asf_welsh_energy_consultation/analysis/produce_plots_and_stats.py --local_data_dir <YOUR_LOCAL_DIR>`. You need to specify the path to the local
```
line 25 needs changing now :)  produce_plots_and_stats instead of produce_plots.py

sofiapinto commented on asf_welsh_energy_consultation/getters/get_data.py:
```diff
@@ -42,7 +41,7 @@ def create_argparser():
+        help='Name of directory where supplementary data is stored in the form `data_YYYYMM`. Defaults to "newest"',
         default="newest",
         type=str,
     )
```
lines below need to be changed to this in order for the code to run. I've basically replaced `/inputs/ `with `/inputs/data `everywhere.

```
def get_args():
    """
    Get arguments from Argument Parser.

    Returns:
        List of arguments.
    """
    parser = create_argparser()

    args = parser.parse_args()

    if args.supp_data == "newest":
        subdirs = [subdir for subdir in os.listdir("inputs/data")]
        args.supp_data = max(subdirs)

    return args


arguments = get_args()

input_data_path = f"inputs/data/{arguments.supp_data}/"

wales_epc_path = "wales_epc.csv"

postcode_path = f"inputs/data/{arguments.supp_data}/postcodes"
regions_path = f"inputs/data/{arguments.supp_data}/regions.csv"
off_gas_path = f"inputs/data/{arguments.supp_data}/off-gas-live-postcodes-2022.xlsx"
oa_path = f"inputs/data/{arguments.supp_data}/postcode_to_output_area.csv"
rurality_path = f"inputs/data/{arguments.supp_data}/rurality.ods"
tenure_path = f"inputs/data/{arguments.supp_data}/tenure.csv"

```

sofiapinto reviewed (commented):
Hey Roisin,

Thanks for the changes and the hard work on improving this! This looks great :)

There are only 3 small things I think you should really change (i've left individual comments for each one of them):
- UPRN instead of BUILDING_REFERENCE_BUBER in correct_new_dwelling_labels() - or let me know if i am looking at this wrong;
- update the data paths to use /inputs/data/ instead of /inputs;
- The labels for figure 4 in welsh are on top of each other; maybe you can rotate them by an angle? or make the font size smaller?

crispy-wonton committed:
allow label rotation on x-axis of bar plots and set for welsh fig 4 to ensure readability

crispy-wonton committed:
fix correct_new_dwelling_label function

- fill NA rank with highest rank
- use UPRN instead of BUILDING_REFERENCE_NUMBER to groupby

crispy-wonton committed:
update directory structure and script name in README

crispy-wonton committed:
fix paths to supplementary datasets and add supplementary data files to config

crispy-wonton committed:
remove 'fig' labels from figures

crispy-wonton committed:
add October 2023 analysis filenames and sources to readme

crispy-wonton committed:
add supp_data arg documentation to README

crispy-wonton committed:
edit get_countries function to work with old and new postcodes datasets
